pkg/address/normalizer.go

```diff
@@ -17,56 +17,64 @@ package address
 import (
 	"errors"
 	"fmt"
-	"os"
 
 	"github.com/go-kit/log"
-	"github.com/google/pprof/profile"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
 
-	"github.com/parca-dev/parca-agent/pkg/objectfile"
+	"github.com/parca-dev/parca-agent/pkg/process"
 )
 
-type ObjectFileCache interface {
-	ObjectFileForProcess(pid int, m *profile.Mapping) (*objectfile.MappedObjectFile, error)
-}
-
 // normalizer is a normalizer that converts memory addresses to position-independent addresses.
 type normalizer struct {
 	logger log.Logger
-
-	objCache ObjectFileCache
+	// normalizationEnabled indicates whether the profiler has to
+	// normalize sampled addresses for PIC/PIE (position independent code/executable).
+	normalizationEnabled  bool
+	normalizationAttempts *prometheus.CounterVec
 }
 
 // NewNormalizer creates a new AddressNormalizer.
-func NewNormalizer(logger log.Logger, objCache ObjectFileCache) *normalizer {
+func NewNormalizer(logger log.Logger, reg prometheus.Registerer, normalizationEnabled bool) *normalizer {
 	return &normalizer{
-		logger:   logger,
-		objCache: objCache,
+		logger:               logger,
+		normalizationEnabled: normalizationEnabled,
+		normalizationAttempts: promauto.With(reg).NewCounterVec(
+			prometheus.CounterOpts{
+				Name:        "parca_agent_profiler_normalization_attempts_total",
+				Help:        "Total number of attempts normalizing frame addresses.",
+				ConstLabels: map[string]string{"type": "cpu"},
+			},
+			[]string{"status"},
+		),
 	}
 }
 
 // Normalize calculates the base addresses of a position-independent binary and normalizes captured locations accordingly.
-func (n *normalizer) Normalize(pid int, m *profile.Mapping, addr uint64) (uint64, error) {
+func (n *normalizer) Normalize(m *process.Mapping, addr uint64) (uint64, error) {
+	if !n.normalizationEnabled {
+		return addr, nil
+	}
+
 	if m == nil {
 		return 0, errors.New("mapping is nil")
 	}
-
-	if m.Unsymbolizable() { // TODO: re-evaluate this condition and return values.
-		return addr, nil
+	if m.Pathname == "" {
+		return 0, errors.New("mapping pathname is empty")
 	}
 
-	objFile, err := n.objCache.ObjectFileForProcess(pid, m)
-	if err != nil {
-		if !(errors.Is(err, objectfile.ErrNoFile) || errors.Is(err, os.ErrNotExist)) {
-			return 0, fmt.Errorf("failed to open obj file: %w", err)
-		}
+	if m.Pathname == "[vdso]" {
```

Comment:
Should we deal with other special mappings? Such as vsyscall, JIT sections, etc?